spec.bs

```diff
@@ -2931,8 +3249,17 @@ The <dfn for=Navigator method>updateAdInterestGroups()</dfn> method steps are:
                   reporting ID=] to it.
             1. [=list/Append=] |igAd| to |ig|'s |interestGroupField|.
 
-        </dl>
+        <dt>"`additionalBidKey`"
+        <dd>
+        1. Let |decodedKey| be the result of running [=forgiving-base64 decode=] with |value|.
+        1. Jump to the step labeled <i><a href=#abort-update>Abort update</a></i> if any of the
+          following conditions hold:
+            * |decodedKey| is a failure;
+            * |decodedKey|'s [=byte sequence/length=] is not 32;
+            * |parsedUpdate|["`ads`"] [=map/exists=].

```

Comment:
I think this also needs to check that ig doesn't have ads?

Reply:
you mean the existing IG stored in DB that's going to be updated? Do you mind pointing me where is that implemented? I only found the check of [the update does not have ads](https://crsrc.org/c/third_party/blink/common/interest_group/interest_group.cc;drc=216468391c57486f50481dacc4cd92c7339f9ac3;l=260)

Reply:
That check in IsValid is part enough, it's called on combination of update + original group from here:
https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/interest_group_storage.cc;drc=eddbcb143c7462e0b8d60e859b96d678ca0c013c;l=2196

Reply:
oh right. thanks.
But for the spec, stored IG (the original group here) will always be valid, right? Since both join and update which modifies stored IG validates the IG before storing it. So it sounds more like an implementation detail that the spec does not need to mention?

Reply:
Well, previous IG can be valid with ads, and update IG can be valid with additionalBidKey (or vice versa), but their combination won't be; and I don't see where it would be rejected in the spec --- I think the "Replace the interest group that has ig’s owner and name in the browser’s interest group set with ig." step would happily the store the now-invalid combination?

Reply:
ah I totally missed this interesting case.
Added the check.
Also fixed a few errors in the spec:
* it kept appending new ads to existing ads list, instead of overwriting the list with the new ads.
* it didn't check the updated IG's size.

Reply:
I think the opposite direction --- setting .ads on something that had additionalBidKey before may still fail to fail? Hmm, doesn't feel like reusing the validity logic from joinIntererestGroup is easy, sadly.

Reply:
Oh right. Fixed by moving the check to a later place to fail if both ads and keys are not null.